### ISSUE TITLE
Partitioned multidimensional arrays (PArray, not just PVector)?

### ISSUE THREAD

**Issue by stevengj** [fverdugo/PartitionedArrays.jl#97]
For parallel stencil computations (e.g. finite-difference methods), it would be nice to have partitioned multidimensional arrays, with support for arbitrary thickness "ghost" overlap regions (so that you could loop over just the interior of each partition and the ghost regions would handle the stencil boundaries).

Not sure whether that is in scope for this package, or if it is something that should be implemented in another package on top of `PVector`?

**Comment by fverdugo** [comment]
Yes, this is in the scope definitively and multidimensional stencil computations are already supported via PVector with
at most a layer of ghosts. For implicit methods, I guess this is what you need since you would need a PVector to represent A*x = b anyway. But I agree that for explicit or matrix free, then a multi dimensional array would be more natural.

```julia
n_per_dir = (10,10,10)
np_per_dir = (2,3,2)
ghost_per_dir = (true,true,true)
periodic_per_dir = (false,true,true)
ranks = LinearIndices(n_per_dir)
index_partition = uniform_partition(ranks,np_per_dir,n_per_dir,ghost_per_dir,periodic_per_dir)
v = pvector(index_partition)
```
The Pvector v is the 1d version of a 10x10x10 multidimensional array split in 2x3x2 parts with ghost in all directions and periodic boundaries (influences the definition of ghost) in the 2nd and 3rd axis.

**Comment by stevengj** [comment]
Note that you might a ghost layer of thickness more than 1 for higher-order finite-difference stencils.

**Comment by fverdugo** [comment]
Generalizing the thickness of the ghost layer should be very easy. I think it is as easy as changing the hardcoded `1` by `ghost` in `start-=1` and `stop+=1`  in this function:

https://github.com/fverdugo/PartitionedArrays.jl/blob/3349755f92f80ec4d04020fcafcb6bb03859fd38/src/p_range.jl#L702

If I am not wrong, this would allow using `ghost_per_dir = (2,2,2)`  in addition to  `ghost_per_dir = (true,true,true)` which would correspond to `ghost_per_dir = (1,1,1)` .  Luckily, this is orthogonal to the PArray thing.

**Comment by ZhangFengshun** [comment]
> Generalizing the thickness of the ghost layer should be very easy. I think it is as easy as changing the hardcoded `1` by `ghost` in `start-=1` and `stop+=1` in this function:
> 
> https://github.com/fverdugo/PartitionedArrays.jl/blob/3349755f92f80ec4d04020fcafcb6bb03859fd38/src/p_range.jl#L702
> 
> If I am not wrong, this would allow using `ghost_per_dir = (2,2,2)` in addition to `ghost_per_dir = (true,true,true)` which would correspond to `ghost_per_dir = (1,1,1)` . Luckily, this is orthogonal to the PArray thing.

Hi~, I changed the "start-=1 and stop+=1" to "start-=2 and stop+=2" in the local_range function. This change works in the uniform_partition function but failed in the consistent! function.
![image](https://github.com/user-attachments/assets/5e554ec5-d841-4f87-90bd-dd1a27f8ad33)
![image](https://github.com/user-attachments/assets/a3f30f0a-0db6-4381-a381-cdeb233481e3)
The pvector of rank 2 should be [1, 1, 2, 2] as expected.

**Comment by ZhangFengshun** [comment]
And I got a AssertionError: is_consistent(graph) when I set np=3 without periodic BC
![image](https://github.com/user-attachments/assets/de182e85-65bf-44cf-939e-021e445cd479)


> > Generalizing the thickness of the ghost layer should be very easy. I think it is as easy as changing the hardcoded `1` by `ghost` in `start-=1` and `stop+=1` in this function:
> > https://github.com/fverdugo/PartitionedArrays.jl/blob/3349755f92f80ec4d04020fcafcb6bb03859fd38/src/p_range.jl#L702
> > 
> > If I am not wrong, this would allow using `ghost_per_dir = (2,2,2)` in addition to `ghost_per_dir = (true,true,true)` which would correspond to `ghost_per_dir = (1,1,1)` . Luckily, this is orthogonal to the PArray thing.
> 
> Hi~, I changed the "start-=1 and stop+=1" to "start-=2 and stop+=2" in the local_range function. This change works in the uniform_partition function but failed in the consistent! function. ![image](https://private-user-images.githubusercontent.com/89072581/356854163-5e554ec5-d841-4f87-90bd-dd1a27f8ad33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMzNTY3OTksIm5iZiI6MTcyMzM1NjQ5OSwicGF0aCI6Ii84OTA3MjU4MS8zNTY4NTQxNjMtNWU1NTRlYzUtZDg0MS00Zjg3LTkwYmQtZGQxYTI3ZjhhZDMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDA2MDgxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3ZmU0MDBhNDQyMTM5ZTRhZDg4MWE3YWQ3YzBjYzdjYzE3MjUyOWIxNjhjYzI5NjZjYzZlMjQ2ZDMxYjI2M2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cbklCzV2ipDoUt1BUZylef5-mKhJa4hb0jZp3Z5aiCA) ![image](https://private-user-images.githubusercontent.com/89072581/356854183-a3f30f0a-0db6-4381-a381-cdeb233481e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMzNTY3OTksIm5iZiI6MTcyMzM1NjQ5OSwicGF0aCI6Ii84OTA3MjU4MS8zNTY4NTQxODMtYTNmMzBmMGEtMGRiNi00MzgxLWEzODEtY2RlYjIzMzQ4MWUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODExVDA2MDgxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhZjJmNzBlYTcxNjhmNzY3M2UxYWZlYWRmOTJjNmRiZTljMDdjMGYwODM0NGJmYjNlODczZDU4MjcxZjYxOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uMNvtrIGue1vazO_N7Cjg3qtNHmH_YpORGD4-xXQG3k) The pvector of rank 2 should be [1, 1, 2, 2] as expected.

**Comment by ZhangFengshun** [comment]
This error is working as expected, and I find that the block_with_constant_size function should also be updated for consistency with the 2 ghost layers.
![image](https://github.com/user-attachments/assets/5690c94a-eb3a-4b63-80b1-d023a9f063ef)

In this way, we got the right consistent! :
![image](https://github.com/user-attachments/assets/6683dc67-fd83-46db-bd75-eb6c64974c31)

**Comment by fverdugo** [comment]
Hi @ZhangFengshun this looks promising, but we do not want to hard code 2 as the thickness of the ghost layer. Instead we want to use the value of variable 'ghost' already available in these functions. Now 'ghost' is assumed to be a bool, but we want to generalise it to an arbitrary integer. Feel free to propose the required changes.

**Comment by fverdugo** [comment]
> For parallel stencil computations (e.g. finite-difference methods), it would be nice to have partitioned multidimensional arrays, with support for arbitrary thickness "ghost" overlap regions (so that you could loop over just the interior of each partition and the ghost regions would handle the stencil boundaries).

Done in #173
